config/src/shortcuts/gesture.rs

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: MPL-2.0
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize};
+
+use super::action::Direction;
+
+/// Description of a gesture that can be handled by the compositor
+#[serde_with::serde_as]
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Serialize, Hash)]
+#[serde(deny_unknown_fields)]
+pub struct Gesture {
+    /// How many fingers are held down
+    pub fingers: u32,
+    pub direction: Direction,
+    // A custom description for a custom binding
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub description: Option<String>,
+}
+
+impl Gesture {
+    /// Creates a new gesture from a number of fingers and a direction
+    pub fn new(fingers: impl Into<u32>, direction: impl Into<Direction>) -> Gesture {
+        Gesture {
+            fingers: fingers.into(),
+            direction: direction.into(),
+            description: None,
+        }
+    }
+
+    /// Append the binding to an existing string
+    pub fn to_string_in_place(&self, string: &mut String) {
+        string.push_str(&format!(
+            "{} Finger {}",
+            self.fingers,
+            self.direction.to_string()
+        ));
+    }
+}
+
+impl ToString for Gesture {
+    fn to_string(&self) -> String {
+        let mut string = String::new();
+        self.to_string_in_place(&mut string);
+        string
+    }
+}
```

Comment:
Clippy recommends implementing `std::fmt::Display` instead, so that it can be used directly in a format arg without extra string allocations.